### PR TITLE
Skip schema compatibility check as admin

### LINKED ISSUE #169: Skip schema compatibility check as admin
As a Galapagos admin, I would like to be able to "skip" the schema compatibility check when a new schema version is released in exceptional cases.

### PR DESCRIPTION
**Pull request by PolatEmre-hg**
closes #169

### CODE REVIEW

**Comment by albrechtflo-hg** [comment]
Currently merge conflicts, please resolve

**Comment by PolatEmre-hg** [comment]
> Currently merge conflicts, please resolve

File: src/main/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImpl.java [hunk 1]
```diff
@@ -443,8 +443,12 @@ public CompletableFuture<SchemaMetadata> addTopicSchemaVersion(String environmen
                 }
 
                 boolean reverse = metadata.getType() == TopicType.COMMANDS;
-                SchemaUtil.verifyCompatibleTo(reverse ? newSchema : previousSchema,
-                        reverse ? previousSchema : newSchema);
+
+                if (!skipCompatCheck) {
```
Comment (albrechtflo-hg) [inline src/main/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImpl.java:447]: Could be included in condition in line 435 - whole block does not need to be executed if `skipCompatCheck` is active.

File: src/test/java/com/hermesworld/ais/galapagos/topics/controller/TopicControllerTest.java [hunk 2]
```diff
@@ -231,7 +231,7 @@ public void testRemoveTopicProducer_notAuthorizedForTopic_negative() {
         ValidatingTopicService topicService = mock(ValidatingTopicService.class);
 
         TopicController controller = new TopicController(topicService, kafkaClusters, applicationsService,
-                namingService);
+                namingService, userService);
 
         // WHEN I am authorized for the producer, but not the topic owning application
         when(applicationsService.isUserAuthorizedFor("app-9")).thenReturn(true);
```
Comment (albrechtflo-hg) [inline src/test/java/com/hermesworld/ais/galapagos/topics/controller/TopicControllerTest.java:237]: Please add two tests, testing the currentUser.isAdmin check (positive and negative case) when the `skipCompatCheck` flag is active.

File: src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java [hunk 3]
```diff
@@ -635,7 +657,7 @@ public void testAddSchemaVersion_withMetadata() throws Exception {
         schema2.setSchemaVersion(2);
         schema2.setChangeDescription("some nice description :)");
 
-        SchemaMetadata newSchemaMetadata = service.addTopicSchemaVersion("test", schema2).get();
+        SchemaMetadata newSchemaMetadata = service.addTopicSchemaVersion("test", schema2, false).get();
```
Comment (albrechtflo-hg) [inline src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java:660]: Parameter should be set to true to have previous behaviour here

File: src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java [hunk 4]
```diff
@@ -662,7 +684,7 @@ public void testAddSchemaVersion_withMetadata_illegalVersionNo_empty() throws Ex
         schema1.setSchemaVersion(2);
 
         try {
-            service.addTopicSchemaVersion("test", schema1).get();
+            service.addTopicSchemaVersion("test", schema1, false).get();
```
Comment (albrechtflo-hg) [inline src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java:687]: Parameter should be set to true to have previous behaviour here

File: src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java [hunk 5]
```diff
@@ -699,7 +721,7 @@ public void testAddSchemaVersion_withMetadata_illegalVersionNo_notMatching() thr
         schema2.setSchemaVersion(3);
 
         try {
-            service.addTopicSchemaVersion("test", schema2).get();
+            service.addTopicSchemaVersion("test", schema2, false).get();
```
Comment (albrechtflo-hg) [inline src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java:724]: Parameter should be set to true to have previous behaviour here

File: src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java [hunk 6]
```diff
@@ -720,7 +742,7 @@ public void testAddSchemaVersion_invalidSchema() throws Exception {
         topicRepository.save(topic1).get();
 
         try {
-            service.addTopicSchemaVersion("test", "topic-1", "{ \"title\": 17 }", null).get();
+            service.addTopicSchemaVersion("test", "topic-1", "{ \"title\": 17 }", null, false).get();
```
Comment (albrechtflo-hg) [inline src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java:745]: Parameter should be set to true to have previous behaviour here

File: src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java [hunk 7]
```diff
@@ -741,7 +763,7 @@ public void testAddSchemaVersion_invalidJson() throws Exception {
         topicRepository.save(topic1).get();
 
         try {
-            service.addTopicSchemaVersion("test", "topic-1", "{", null).get();
+            service.addTopicSchemaVersion("test", "topic-1", "{", null, false).get();
```
Comment (albrechtflo-hg) [inline src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java:766]: Parameter should be set to true to have previous behaviour here

File: src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java [hunk 8]
```diff
@@ -766,7 +788,7 @@ public void testAddSchemaVersion_DataObjectSimpleAtJSONSchema() throws Exception
         topicRepository.save(topic1).get();
 
         try {
-            service.addTopicSchemaVersion("test", "topic-1", testJsonSchema, null).get();
+            service.addTopicSchemaVersion("test", "topic-1", testJsonSchema, null, false).get();
```
Comment (albrechtflo-hg) [inline src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java:791]: Parameter should be set to true to have previous behaviour here

File: src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java [hunk 9]
```diff
@@ -790,7 +812,7 @@ public void testAddSchemaVersion_DataObjectNestedAtJSONSchema() throws Exception
 
         topicRepository.save(topic1).get();
 
-        service.addTopicSchemaVersion("test", "topic-1", testJsonSchema, null).get();
+        service.addTopicSchemaVersion("test", "topic-1", testJsonSchema, null, false).get();
```
Comment (albrechtflo-hg) [inline src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java:815]: Parameter should be set to true to have previous behaviour here

File: src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java [hunk 10]
```diff
@@ -810,7 +832,7 @@ public void testAddSchemaVersion_NoSchemaProp() throws Exception {
         topicRepository.save(topic1).get();
 
         try {
-            service.addTopicSchemaVersion("test", "topic-1", testJsonSchema, null).get();
+            service.addTopicSchemaVersion("test", "topic-1", testJsonSchema, null, false).get();
```
Comment (albrechtflo-hg) [inline src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java:835]: Parameter should be set to true to have previous behaviour here

File: src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java [hunk 11]
```diff
@@ -1065,7 +1087,7 @@ public void testAddSchemaVersion_DataObjectNestedAtJSONSchemaAndDataTopic() thro
 
         topicRepository.save(topic1).get();
 
-        service.addTopicSchemaVersion("test", "topic-1", testJsonSchema, null).get();
+        service.addTopicSchemaVersion("test", "topic-1", testJsonSchema, null, false).get();
```
Comment (albrechtflo-hg) [inline src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java:1090]: Parameter should be set to true to have previous behaviour here

File: src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java [hunk 12]
```diff
@@ -1096,7 +1118,7 @@ public void testAddSchemaVersion_WithChangeDesc() throws Exception {
         newSchema.setSchemaVersion(2);
         newSchema.setChangeDescription("Added new schema which is better");
 
-        service.addTopicSchemaVersion("test", newSchema).get();
+        service.addTopicSchemaVersion("test", newSchema, false).get();
```
Comment (albrechtflo-hg) [inline src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java:1121]: Parameter should be set to true to have previous behaviour here

File: src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java [hunk 13]
```diff
@@ -1125,7 +1147,7 @@ public void testAddSchemaVersion_WithChangeDesc_negative() throws Exception {
         newSchema.setChangeDescription("Added new schema which is better");
 
         try {
-            service.addTopicSchemaVersion("test", newSchema).get();
+            service.addTopicSchemaVersion("test", newSchema, false).get();
```
Comment (albrechtflo-hg) [inline src/test/java/com/hermesworld/ais/galapagos/topics/service/impl/TopicServiceImplTest.java:1150]: Parameter should be set to true to have previous behaviour here

File: ui/src/app/layout/createtopic/createtopic.component.ts [hunk 14]
```diff
@@ -139,7 +139,7 @@ export class CreateTopicComponent implements OnInit {
     }
 
     private async createInitialSchema(topicName: string, environmentId: string, jsonSchema: string): Promise<any> {
-        return this.topicsSerivce.addTopicSchema(topicName, environmentId, jsonSchema).then(
+        return this.topicsSerivce.addTopicSchema(topicName, environmentId, jsonSchema, false).then(
```
Comment (albrechtflo-hg) [inline ui/src/app/layout/createtopic/createtopic.component.ts:142]: Won't work if current user is not admin?!

File: ui/src/app/layout/topics/schemasection/schema-section.component.html [hunk 15]
```diff
@@ -53,6 +53,12 @@
                             <input [(ngModel)]="schemaChangeDescription" type="text" size="50"
                                    class="form-control"/><br>
                         </div>
+                        <div *ngIf="isAdmin" class="checkbox">
+                            <label>
+                                <input type="checkbox"
+                                       [(ngModel)]="skipCompatCheck"> {{ 'Skip compatibility check' | translate }}
```
Comment (albrechtflo-hg) [inline ui/src/app/layout/topics/schemasection/schema-section.component.html:59]: An `ngb-alert` with warning type should be included here and state, "**Caution**: This allows for incompatible changes in the topic data. Only use this if you really know what you are doing."

The alert should only be displayed when the checkbox is checked.

**Comment by albrechtflo-hg** [comment]
Ignore all my "parameter must be true for previous behaviour" comments. Of course, "false" is correct, I misthought it for being "check yes / no" (instead of "skip check yes/no").

File: ui/src/assets/i18n/de.json [hunk 16]
```diff
@@ -290,6 +290,8 @@
     "EVENTS": "Events",
     "DATA": "Master Data",
     "COMMANDS": "Commands",
-    "INTERNAL": "Anwendungsintern",
     "SAVE_DEV_KEY_WARNING": "Bitte <b>speichere</b> das Secret, bevor du diese Seite verlässt! Das Secret wird nicht in Galapagos gespeichert. Bitte beachte auch, dass der API Key folgendes <b>Verfallsdatum</b> besitzt: {{ expiresAt }}"
+    "INTERNAL": "Anwendungsintern",
+    "Skip compatibility check": "Schema-Kompatibilitätsprüfung überspringen",
+    "SKIP_COMBAT_CHECK_WARNING": "<b>Achtung</b>: Dies ermöglicht inkompatible Änderungen in den Topic Daten. Verwende dies nur, wenn du wirklich weißt, was du tust."
```
Comment (albrechtflo-hg) [inline ui/src/assets/i18n/de.json:296]: No, no combat please, stop war! ;-)  (Please rename key to SKIP_COMPAT_...)

**Review by albrechtflo-hg — CHANGES_REQUESTED** [review]
Please rename SKIP_COMBAT_... to SKIP_COMPAT

**Review by albrechtflo-hg — APPROVED** [review]

**Pull request merged**